src/docfx/lib/msalias/MicrosoftAliasCache.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class MicrosoftAliasCache : IDisposable
+    {
+        private readonly string _cachePath;
+        private readonly double _expirationInHours;
+        private readonly SemaphoreSlim _syncRoot = new SemaphoreSlim(1, 1);
+        private readonly MicrosoftGraphAccessor _microsoftGraphAccessor;
+        private Dictionary<string, MicrosoftAlias> _aliases = new Dictionary<string, MicrosoftAlias>();
+        private IList<string> _tempInvalidAliases = new List<string>();
+        private bool _needUpdate = false;
+
+        public MicrosoftAliasCache(Config config)
+        {
+            _cachePath = AppData.MicrosoftAliasCachePath;
+            _expirationInHours = config.JsonSchema.MicrosoftAliasCacheExpirationInHours;
+
+            _microsoftGraphAccessor = new MicrosoftGraphAccessor(
+                config.JsonSchema.MicrosoftGraphTenantId,
+                config.JsonSchema.MicrosoftGraphClientId,
+                config.JsonSchema.MicrosoftGraphClientSecret);
+
+            if (File.Exists(_cachePath))
+            {
+                var cacheFile = JsonUtility.Deserialize<MicrosoftAliasCacheFile>(ProcessUtility.ReadFile(_cachePath), _cachePath);
+                _aliases = cacheFile.Aliases.Where(msAlias => msAlias.Expiry >= DateTime.UtcNow).ToDictionary(msAlias => msAlias.Alias);
+            }
+        }
+
+        public Task<(Error error, MicrosoftAlias msAlias)> GetAsync(string alias)
+        {
+            if (string.IsNullOrEmpty(alias))
```

Comment:
When tenantId, or clientId, or secret is missing, we cannot give the correct validation result because we cannot call msgraph API. In this case, return here and don't put it to cache. Can init `_msGraphAccessor` to `null` when any of the above configs are missing.